pkg/oci/build_metadata.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/cilium/ebpf"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/types"
+)
+
+func loadSpec(progContent []byte) (*ebpf.CollectionSpec, error) {
+	progReader := bytes.NewReader(progContent)
+	spec, err := ebpf.LoadCollectionSpecFromReader(progReader)
+	if err != nil {
+		return nil, fmt.Errorf("loading spec: %w", err)
+	}
+	return spec, err
+}
+
+func getAmd64Spec(opts *BuildGadgetImageOpts) (*ebpf.CollectionSpec, error) {
+	// TODO: we could perform a sanity check to be sure different architectures are the same,
+	// but it's too much
+	progPath, ok := opts.EBPFObjectPaths[ArchAmd64]
+	if !ok {
+		return nil, fmt.Errorf("missing eBPF object for %s", ArchAmd64)
+	}
+
+	progContent, err := os.ReadFile(progPath)
+	if err != nil {
+		return nil, fmt.Errorf("reading eBPF object file: %w", err)
+	}
+
+	return loadSpec(progContent)
+}
```

Comment:
The code @eiffel-fl gave seems good. We can extend it to try to get the `ArchArm64` spec if it fails for `amd64`. This could happen if the gadget only supports `arm64` itself (For whatever reason)

Reply:
> Why is there only one architecture?

We only need to get a single file to get generate the metadata from it. I changed the code to try with all available architectures and made a comment to make it clear it's only used for generating the metadata and doesn't mean we're only implementing support for a single architecture.